src/workerd/server/server-test.c++

```diff
@@ -199,6 +237,73 @@ private:
     buffer.add('\0');
     return kj::String(buffer.releaseAsArray());
   }
+
+  kj::String readWebSocketMessage(size_t maxMessageSize = 1 << 24) {
+    // Reads a single, non-fragmented WebSocket message. Returns just the payload.
+    kj::Vector<uint8_t> header(256);
+    kj::Vector<uint8_t> mask(4);
+
+    KJ_IF_MAYBE(p, premature) {
+      header.add(*p);
+      premature = kj::Maybe<char>();
+    }
+
+    tryRead(header, 2 - header.size(), "reading first two bytes of header");
+    bool masked = header[1] & 0x80;
+    size_t sevenBitPayloadLength = header[1] & 0x7f;
+    size_t realPayloadLength = sevenBitPayloadLength;
+
+    if (sevenBitPayloadLength == 126) {
+      tryRead(header, 2, "reading 16-bit payload length");
+      realPayloadLength = (static_cast<size_t>(header[2]) << 8) + static_cast<size_t>(header[3]);
+    } else if (sevenBitPayloadLength == 127) {
+      tryRead(header, 8, "reading 64-bit payload length");
+      realPayloadLength = (static_cast<size_t>(header[2]) << 56)
+        + (static_cast<size_t>(header[3]) << 48)
+        + (static_cast<size_t>(header[4]) << 40)
+        + (static_cast<size_t>(header[5]) << 32)
+        + (static_cast<size_t>(header[6]) << 24)
+        + (static_cast<size_t>(header[7]) << 16)
+        + (static_cast<size_t>(header[8]) << 8)
+        + (static_cast<size_t>(header[9]));
+
+      if (realPayloadLength > maxMessageSize) {
```

Comment:
Non-blocking nit: this could use `KJ_REQUIRE(realPayloadLength <= maxMessageSize, ...` for a minor simplification.

Reply:
Done.